substratevm/src/com.oracle.svm.reflect/src/com/oracle/svm/reflect/serialize/hosted/SerializationFeature.java

```diff
@@ -205,6 +227,63 @@ private void abortIfSealed() {
         UserError.guarantee(!sealed, "Too late to add classes for serialization. Registration must happen in a Feature before the analysis has finished.");
     }
 
+    @Override
+    public void registerAllAssociatedClasses(ConfigurationCondition condition, Class<?> clazz) {
+        registerAllAssociatedClasses(condition, clazz, new HashSet<>());
+    }
+
+    private void registerAllAssociatedClasses(ConfigurationCondition condition, Class<?> clazz, Set<Class<?>> alreadyVisited) {
+        if (alreadyVisited.contains(clazz)) {
+            return;
+        }
+        alreadyVisited.add(clazz);
+        String targetClassName = clazz.getName();
+        // If the serialization target is primitive, it needs to get boxed, because the target is
+        // always an Object.
+        if (clazz.isPrimitive()) {
+            Class<?> boxedType = JavaKind.fromJavaClass(clazz).toBoxedJavaClass();
+            registerAllAssociatedClasses(condition, boxedType, alreadyVisited);
+            return;
+        } else if (!Serializable.class.isAssignableFrom(clazz)) {
+            println("Warning: The given class " + targetClassName + " is not serializable, it wouldn't get registered.\n");
+            return;
+        } else if (access.findSubclasses(clazz).size() > 1) {
+            println("Warning: The given class " + targetClassName +
+                            " has subclasses. We can't figure out which one should be registered. Skip.\n");
+            return;
+        }
+        try {
+            clazz.getDeclaredMethod("writeObject", ObjectOutputStream.class);
+            println("Warning: The given class " + targetClassName +
+                            " has customized serialization method writeObject. " +
+                            "We can't figure out what should be registered. Skip.\n");
+            return;
+        } catch (NoSuchMethodException e) {
+            // Expected case. Do nothing
```

Comment:
Aside from the `println` call which prints to `System.out` instead of `System.err` and could be renamed to `warn` as that's what it is used for (but I blame not having a proper facility for that anywhere), the error messages can be improved:
`Class foo.bar does not implement java.io.Serializable and was not registered for object serialization.`
`Class foo.bar has more than one subclass. No subclasses were registered for object serialization.`
`Class foo.bar implements its own writeObject method for object serialization. Any serialization types it uses need to be explicitly registered.`

Reply:
Fixed as suggested. But for the subclass case, we set the checking threshold to 1 because the `access.findSubclasses` API returns a list of subclasses including the base class itself. We don't registered the base class and any of its subclasses in this case.